api/file.go

```diff
@@ -103,82 +105,92 @@ func uploadFile(c *Context, w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	files := m.File["files"]
+	if !HasPermissionToChannelContext(c, channelId, model.PERMISSION_UPLOAD_FILE) {
+		return
+	}
 
 	resStruct := &model.FileUploadResponse{
-		Filenames: []string{},
+		FileInfos: []*model.FileInfo{},
 		ClientIds: []string{},
 	}
 
-	imageNameList := []string{}
+	previewPathList := []string{}
+	thumbnailPathList := []string{}
 	imageDataList := [][]byte{}
 
-	if !HasPermissionToChannelContext(c, channelId, model.PERMISSION_UPLOAD_FILE) {
-		return
-	}
-
-	for i := range files {
-		file, err := files[i].Open()
+	for i, fileHeader := range m.File["files"] {
+		file, fileErr := fileHeader.Open()
 		defer file.Close()
-		if err != nil {
-			http.Error(w, err.Error(), http.StatusInternalServerError)
+		if fileErr != nil {
+			http.Error(w, fileErr.Error(), http.StatusInternalServerError)
 			return
 		}
 
 		buf := bytes.NewBuffer(nil)
 		io.Copy(buf, file)
 
-		filename := filepath.Base(files[i].Filename)
+		info, err := model.GetInfoForBytes(fileHeader.Filename, buf.Bytes())
+		if err != nil {
+			c.Err = err
+			c.Err.StatusCode = http.StatusBadRequest
+			return
+		}
+
+		info.Id = model.NewId()
+		info.UserId = c.Session.UserId
 
-		uid := model.NewId()
+		filename := filepath.Base(fileHeader.Filename)
+		pathPrefix := "teams/" + c.TeamId + "/channels/" + channelId + "/users/" + c.Session.UserId + "/" + info.Id + "/"
```

Comment:
True, it just seems strange to store them like that unnecessarily.  Discuss in group review.